### DISCUSSION TITLE
Admin create self service recovery link cause error

### DISCUSSION THREAD

**Discussion by thongdvsmartosc** [ory/kratos#2812] [Q&A] [answered]
Error Response:
`&{403 Forbidden 403 HTTP/1.1 1 1 map[Content-Length:[843] Content-Type:[application/json] Date:[Mon, 17 Oct 2022 03:05:13 GMT] Server:[envoy] Set-Cookie:[csrf_token_446062ac90ef3c5ab45e5fdc57a424d0b252ff28143597f80151dd915ea9f5b5=JGwV3HP+F+ZviVkFnP6MKZJuz0MuSaYV/9dtSirxYmY=; Path=/; Domain=kincloud.io; Max-Age=31536000; HttpOnly; Secure; SameSite=Lax] Vary:[Cookie] X-Envoy-Upstream-Service-Time:[1]] {{"error":{"id":"security_csrf_violation","code":403,"status":"Forbidden","request":"05cbc4f9-c7dc-49ed-a2d1-db3c9970f44c","reason":"Please retry the flow and optionally clear your cookies. The request was rejected to protect you from Cross-Site-Request-Forgery (CSRF) which could cause account takeover, leaking personal information, and other serious security issues.","details":{"docs":"https://www.ory.sh/kratos/docs/debug/csrf","hint":"The anti-CSRF cookie was found but the CSRF token was not included in the HTTP request body (csrf_token) nor in the HTTP Header (X-CSRF-Token).","reject_reason":"The HTTP Cookie Header was set and a CSRF token was sent but they do not match. We recommend deleting all cookies for this domain and retrying the flow."},"message":"the request was rejected to protect you from Cross-Site-Request-Forgery"}}
} 843 [] false false map[] 0xc00051e300 0xc0009fe0b0}`

source code:
`

        client := pkg.NewSDKForSelfHosted("https://mydomain.com/api/kratos/public")
	ctx := context.Background()

	flow, res, err := client.V0alpha2Api.InitializeSelfServiceRecoveryFlowWithoutBrowser(ctx).Execute()

	result, res, err := client.V0alpha2Api.AdminCreateSelfServiceRecoveryLink(ctx).AdminCreateSelfServiceRecoveryLinkBody(
		ory.AdminCreateSelfServiceRecoveryLinkBody{
			IdentityId: "identity id",
			ExpiresIn:  ory.PtrString("24h"),
		},
	).Execute()
`

version: v0.9.0-alpha.3

dsn: "this is set in environment var DSN"

serve:
  public:
    base_url: xxx
    request_log:
      disable_for_health: true
    cors:
      enabled: false
  admin:
    base_url: xxx
    request_log:
      disable_for_health: true

selfservice:
  default_browser_return_url: xxx
  allowed_return_urls:
  - xxx
  - ...
  methods:
    link:
      enabled: true
      config:
        lifespan: 24h
    password:
      enabled: true
  flows:
    recovery:
      enabled: true
      ui_url: xxx

    settings:
      ui_url: xxx
      privileged_session_max_age: 24h

    error:
      ui_url: xxx

    logout:
      after:
        default_browser_return_url: xxx
    
    login:
      ui_url: xxx
      lifespan: 10m
    
    registration:
      lifespan: 10m
      ui_url: xxx
      after:
        password:
          hooks:
          - hook: web_hook
            config:
              url: xxx
              method: POST
              body: xxx
              auth:
                type: api_key
                config:
                  name: xxx
                  value: xxx
                  in: header
          - hook: web_hook
            config:
              url: xxx
              method: POST
              body: file:///etc/kratos/post_register_hook.jsonnet
              auth:
                type: api_key
                config:
                  name: xxx
                  value: xxx
                  in: header

log:
  level: debug
  format: json
  leak_sensitive_values: true

secrets:
  cookie:
  - xxx

hashers:
  algorithm: bcrypt
  bcrypt:
    cost: 8
  argon2:
    parallelism: 1
    memory: 128MB
    iterations: 2
    salt_length: 16
    key_length: 16

identity:
  default_schema_id: default
  schemas:
    - id: default
      url: file:///etc/kratos/identity.schema.json
    - id: tenant
      url: file:///etc/kratos/identity.schema.json

cookies:
  domain: xxx

session:
  cookie:
    domain: xxx

courier:
  smtp:
    connection_uri: xxx
    from_address: xxx

I have a self hosted kratos. I cannot create a recovery link. Can you explain why, please? Thank you.

**Comment by vinckr** [comment, marked as answer]
Hello @thongdvsmartosc, 
Could you please share a reproducible repository with shell commands (e.g. kratos identities, curl API requests...) that reproduces this issue? How are you hosting Ory Kratos?

**Reply by MUzairS15** [reply]
What’s the solution for this? When trying to do password recovery via admin using sdk it fails giving above mentioned error. Were as a post request to ‘/admin/recovery/link’ gives recovery link as response but user doesn’t receive any email with the link

**Reply by vinckr** [reply]
Hello @MUzairS15 

As stated above, just knowing that something is not working as expected gives us no clue as to what you are doing, how you are using Ory software, what you want to achieve, etc. 

We need more information to be able to help you. 

How are you running Ory Kratos and what version are you on?  
Please make sure you are on the latest version, so you are not running into bugs that are already fixed, thanks 🙏 

(I will mark this issue as answered since there has been no response from OP for some time, but that does not mean I dont see answers in here. Feel free to respond here if you are facing the same issue, but please make sure to include as many details as possible, so you can be helped promptly.)
